### PR TITLE
Run jenkins test on couchdb2.x

### LINKED ISSUE #47: Unit tests fail on Node.js v9.8.0
* `cloudant-follow` version: `0.17.0-SNAPSHOT`
* Node.js version: `v9.8.0`

Here is what I do to test `cloudant-follow`:

1) `git clone git@github.com:cloudant-labs/cloudant-follow.git`
2) `cd cloudant-follow`
3) `npm install`
4) `cd test`
5) `node couch.js`
This one succeeds.
6) `node follow.js`
This one fails.
7) `node issues.js`
This one fails.
8) `node params.js`
This one fails.
9) `node stream.js`
This one fails.

I suspect the Node.js version to cause this because when I run this in my Node.js program:

```
feed.on("change", function (change) {
    console.log("Change detected:", change)
    this.stop()
})
```

the `this.stop()` line will cause this error and will terminate the program:

```
_http_client.js:354
  if (req.res && req.res.readable) {
          ^

TypeError: Cannot read property 'res' of null
    at Socket.socketCloseListener (_http_client.js:354:11)
    at Socket.emit (events.js:180:13)
    at TCP._handle.close [as _onclose] (net.js:538:12)
```

But maybe the Node.js version is not the source of the problem. I will try to run the tests on another Node.js version tomorrow and I will post the results here.

### LINKED ISSUE #50: Unit tests fail on CouchDB v2.1.1
* `cloudant-follow` version: `0.17.0-SNAPSHOT`

Unit tests fail both with CouchDB v2.1.1 and Node.js v9.8.0. Only the combination of CouchDB v1.7.1 and Node.js v8.9.1 passes all the tests:

image

Related issue: https://github.com/cloudant-labs/cloudant-follow/issues/47

### PR DESCRIPTION
**Pull request by smithsz**
Thanks for your hard work, please ensure all items are complete before opening.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://github.com/cloudant-labs/cloudant-follow/blob/master/DCO1.1.txt)
- [x] You have added tests for any code changes
- [x] You have updated the [CHANGES.md](https://github.com/cloudant-labs/cloudant-follow/blob/master/CHANGES.md)
- [x] You have completed the PR template below:

## What
- Fix Jenkins tests by getting the docker sidecar run to default to docker-in-docker.
- Make unit tests compatible with CouchDB 2.X.
- Remove redundant `agentRemove` socket call (fixes issue in Node.js v9.8.0).
- Run Jenkins tests against Couchdb versions "1.7.1" and "latest".
- Update engines in preparation for Node.js 4 EOL

## Test Coverage

| Node.js Version  | CouchDB@1.7.1 | Latest CouchDB@<i></i>2.X.X  |
| -----------------|:-------------:| --------------:|
| 6                           |||
| 8                           |||
| _latest_                 |||

## Issues
Fixes #47.
Fixes #50.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -15,25 +15,29 @@
...
-          sh """
-            [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"
-            nvm install ${version}
-            nvm use ${version}
-            npm test && npm run unreliable-feed-test
-          """
+    // To get the docker sidecar run to default to docker-in-docker we must
+    // unset the DOCKER_HOST variable.
+    withEnv(["DOCKER_HOST="]){
+      // Install CouchDB 1.6.1
```
Comment (ricellis) [inline Jenkinsfile]: This comment is wrong. It was a bit wrong before when it was installing 1.7.1, but it is very wrong now :)

File: Jenkinsfile [hunk 2]
```diff
@@ -15,25 +15,29 @@
...
+      docker.image('apache/couchdb:latest').withRun('-p 5984:5984') {
```
Comment (ricellis) [inline Jenkinsfile]: Should we fix this at `2.1.1` it's a shame there isn't a `2.x` or similar tag. I'm fine with `latest`, but we have to keep an eye out for CouchDB 3 breaking things. Maybe this fork won't need to live that long(!)

File: test/couch.js [hunk 3]
```diff
@@ -169,6 +172,32 @@ function make_data(minimum_size, callback) {
...
 }
 
+function get_update_seq(seq) {
+  switch (typeof seq) {
+    case 'number':
+      // CouchDB 1.X format, i.e. { "seq":1, ...
+      return seq;
+    case 'string':
+      // CouchDB 2.X format, i.e. { "seq":"1-xxxxxxxx", ...
+      return parseInt(seq.split('-')[0], 10);
```
Comment (ricellis) [inline test/couch.js]: well it's better than failing tests :)

File: test/issues.js [hunk 4]
```diff
@@ -21,36 +21,46 @@ var lib = require('../lib')
...
-    feed.on('stop', function() {
-      // Test using since=-1 (AKA since="now").
-      follow({'db':couch.DB, since:'now'}, function(er, change) {
-        t.equal(change.seq, 4, 'Only get changes made after starting the feed')
-        t.equal(change.id, "You're in now, now", 'Got the subsequent change')
-
-        this.stop()
-        t.end()
-      })
+  // CouchDB 1.X only test
```
Comment (ricellis) [inline test/issues.js]: It'd be good to know why this is a 1.x only test

File: test/stream.js [hunk 5]
```diff
@@ -49,74 +64,82 @@ test('Writatable Stream API', function(t) {
 })
 
 test('Error conditions', function(t) {
-  var feed = new follow.Changes
-  t.throws(write, 'Throw if the feed type is not defined')
+  // CouchDB 1.X only test
```
Comment (ricellis) [inline test/stream.js]: Again why is this a 1.x only test?

File: Jenkinsfile (hunk 2 above)
Comment (cluxter) [reply Jenkinsfile]: There is actually:

image

Source: https://hub.docker.com/_/couchdb/

Comment (smithsz) [reply Jenkinsfile]: Good spot! I'll switch to `apache/couchdb:2`.

File: test/issues.js (hunk 4 above)
Comment (smithsz) [reply test/issues.js]: Add comment.

File: test/stream.js (hunk 5 above)
Comment (smithsz) [reply test/stream.js]: After a bit of jiggery-pokery this now passes.

File: Jenkinsfile (hunk 2 above)
Comment (ricellis) [reply Jenkinsfile]: nice, it's a shame they aren't listed [here](https://github.com/apache/couchdb-docker#available-tags)

**Pull request merged**